Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp

```diff
@@ -3560,6 +3560,19 @@ IntRect WebGLRenderingContextBase::getImageDataSize(ImageData* pixels)
     return texImageSourceSize(*pixels);
 }
 
+void WebGLRenderingContextBase::postProcessPixelBuffer() const
+{
+    if (!canvasBase().shouldInjectNoiseBeforeReadback())
+        return;
+
+    ImageBuffer* buffer = canvasBase().buffer();
+    if (!buffer)
+        return;
+
+    auto canvasRect = texImageSourceSize(canvasBase());
```

Comment:
This function is incorrectly used. E.g. if you have a canvasBase, you know how to get its size without any additional function.
texImageSourceSize function is intended to be used for TexImage sources  for arbitrary source T. The canvas here isn't used as texImageSource, it's just this canvas.

As you can observe, there's nothing webgl about this function. The logic of this function is:
   Post process each pixel that changed after previous post process.

This applies to any canvas.

So one option would be that you would make following chagnes:
- Move CanvasRenderingContext::postProcessPixelBuffer() to CanvasBase::postProcessPixelBuffer()
- Rename  CanvasBase::postProcessPixelBuffer() to  CanvasBase::postProcessRenderingResults()
- Move 
 CanvasRenderingContext::postProcessPixelBuffer(ImageBuffer&, IntRect, const HashSet<uint32_t>&) to CanvasBase::postProcessPixelBuffer(ImageBuffer&, IntRect, const HashSet<uint32_t>&)
- Introduce CanvasBase::didDraw(const std::optional<FloatRect>& rect, SomeOptions) where you record m_postProcessDirtyRect
- Remove m_wasLastDrawPutImageData, this is superseeded by m_postProcessDirtyRect. Current m_wasLastDrawPutImageData doesn't seem to handle putImageData correctly, it will allow attacker to circumvent the post processing by calling with for example 1,1 region as per my reading of my checkout, I may be wrong though
- Make SomeOption to be able to get "SomeOption::TheDrawWasActuallyPutImageDataSoDoNotAccumulateThisToPostProcessingDirtyRect"



So in other words:
- canvas drawing contexts signal CanvasBase that "something changed, maintain the post processing state"
- CanvasBase already has the image buffer for the display buffer  of the context. CanvasBase is already the sole entity giving the pixels out, with the exception of WebGL non-premul alpha PixelBuffer case. Make CanvasBase apply the post processing.

Reply:
Thanks for the review! I agree that your suggestion is better.